Query ISNI database when adding authors

hughrun opened the pull request:
~~This is not ready for prime time because it doesn't actually do anything useful yet.~~

Speculatively may help with #1119 

## Problem statement

Due to various factors Bookwyrm, like most systems dealing with bibliographic data, needs to both disambiguate and deduplicate author records. Lucky for us, that's why the [International Standard Name Identifier](https://isni.org/page/what-is-isni) was created. In even better news, `Author` records already have an `isni` field 🎉 

We can query the ISNI database with a simple `GET` request using keyword name search. The great thing is this works for any recognised variation of an author's name, which doesn't make it foolproof but does give us a high chance of finding a relevant record and therefore avoiding unnecessarily creating a new one.

## Proposed approach

- [X] query ISNI when a book is added
- [X] display potential matches in confirmation screen with description and link to ISNI record
- [x] merge in ISNI and local database author matches (i.e. where we identify that it is the same author, probably based on the ISNI itself)
- [ ] ~~highlight in some kind of encouraging way if there is an ISNI match that is also an existing author (i.e. the above) because this is more likely than other options to be the match we're looking for~~
- [x] display a link to the ISNI page for authors on their author page, since we have an ISNI field anyway
- [x] use the ISNI author description to populate the author.bio if it is otherwise blank
- [x] display with the `nameTitle` (for ISNI records) or a succinct version of the `bio` (local database records) to aid in identifying whether the record is the one you want
- [x] populate "aliases" for new authors automatically from their ISNI record
- [ ] make a command in `./bw-dev` that server admins can run periodically (say once a month using a cron job) to refresh any author records that have an ISNI recorded, if new data (e.g. a new alias) is available.

## Examples

Which "John Smith"?

![smith](https://user-images.githubusercontent.com/17669239/139381238-a0bf2d63-92ae-431e-8a1b-5dca16dae756.png)

John Birmingham the British-born Australian author and John Birmingham the author of _He died with a Felafel in his hand_ are the same person, so these two records should be merged, at least within the "edit book" view.

![existing](https://user-images.githubusercontent.com/17669239/139381558-15d67bc2-067d-4259-8e98-8e4bcaba68d6.png)

What happens if we search for John's name in a totally different alphabet?

![russian](https://user-images.githubusercontent.com/17669239/139381684-dea8d7b2-38f0-4087-bca3-e874ccbf6557.png)

Works in Japanese too:

![tanizaki](https://user-images.githubusercontent.com/17669239/139381707-e4cf9a97-d881-4368-876e-0e60f1348be8.png)

But it doesn't _always_ work the way we would like:

Not great but ok

![søren](https://user-images.githubusercontent.com/17669239/139381809-d921ebac-8001-44c9-a141-a0a663da36a6.png)

Fine I guess

![soeren](https://user-images.githubusercontent.com/17669239/139381818-85ccf973-f3d8-43fb-a9bb-977137b24847.png)

Not so great

![soren](https://user-images.githubusercontent.com/17669239/139381821-56766f56-be5c-4f33-a6ab-0b4218e24700.png)

hughrun commented:
Adding a background tasks will not be included in this PR because (a) it's slightly out of scope and (b) I don't know how to add things to the task manager 😆 

## This PR

1. Queries the open ISNI database when an author is added to a book
2. Suggests relevant authors fetched from ISNI in addition to authors in the local DB
3. Merges the author data in the UI if a local author ISNI ID matches an "ISNI author"
4. Augments local author metadata from ISNI if they match and the local record is incomplete
5. Populates new authors with data from ISNI when adding a new ISNI author

## Caveats

1. Deduping authors across decentralised databases is one of the hard problems of library science. It is basically impossible to have perfect deduplication across all Bookwyrm sites, this PR simply attempts to make it easier to keep duplicates to a minimum.
2. If an existing author does not have an `isni` value, there's not much we can do to deduplicate. It is too dangerous to guess whether we have a match.
3. The Wikipedia link from ISNI is the first one we find, with no guarantees of which language version of Wikipedia it is. I will probably put in an enhancement Issue related to this shortly.

## Note on aliases

Aliases are already checked when searching for a possible match for an author. This PR helps with this by fetching (and deduping) aliases from ISNI. However, aliases are not always recognised. This appears to be mostly or possibly _only_ an issue with Chinese and Japanese scripts. This is not a RTL problem, because Arabic and Korean scripts seem to be fine.

## Note on editing authors

Currently if an author record is manually edited the ISNI and VIAF records are wiped out. #1584 has been logged for this.

## Screenshots and examples

### Author record before adding a new book:

![guido_before](https://user-images.githubusercontent.com/17669239/139650764-ece47873-bbf3-41eb-9f48-1e0eea3e06f1.png)

### Augmented record after adding a new book:

![guido_after](https://user-images.githubusercontent.com/17669239/139651382-2c0eedec-b2b4-4e73-b004-2b64fc21a82e.png)

### Searching for an alias shows both the local information (related book) and ISNI description:

![guido_merged](https://user-images.githubusercontent.com/17669239/139651470-71fd4f6e-283b-4b28-992d-dd613e520636.png)

### Works on any alias:

![wellesley](https://user-images.githubusercontent.com/17669239/139652246-ae7fdb8d-fb9a-4ac3-9b86-80f5b6ba75cb.png)

### Except when it doesn't:

![guido_fail](https://user-images.githubusercontent.com/17669239/139651691-1be313a9-c1b2-42d3-828f-9d397a583f39.png)

### If the local record doesn't have an ISNI there's nothing we can do to merge:
 
![scott_duplicate](https://user-images.githubusercontent.com/17669239/139652302-3851df9d-044e-4157-8f23-3fa9a365ec4d.png)

mouse-reeve commented on bookwyrm/templates/book/edit/edit_book.html:
```diff
@@ -52,19 +53,34 @@ <h2 class="title is-4">{% trans "Confirm Book Info" %}</h2>
+                            {{ isni_match.name }}
+                        </label>
+                        <p class="help ml-5 mb-2">
+                            <a href="{{ isni_match.uri }}" target="_blank">{{ isni_match.bio }}</a>
```
For security, this link should have ` rel="noopener noreferrer"` since it opens an external link

mouse-reeve commented on bookwyrm/utils/isni.py:
```diff
@@ -0,0 +1,152 @@
+    """replace spaces for url encoding"""
+
+    # TODO: this is very lazy and incomplete
+    return string.replace(" ", "+")
```
Is `urllib.parse.quote_plus` the ticket here? (https://docs.python.org/3/library/urllib.parse.html#urllib.parse.quote_plus)

mouse-reeve commented on bookwyrm/utils/isni.py:
```diff
@@ -0,0 +1,152 @@
+        "&maximumRecords=",
+        str(max_records),
+        "&startRecord=1&recordPacking=xml&sortKeys=RLV%2Cpica%2C0%2C%2C",
+    ]
```
[There's an easier way!](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode) You can generate the url by passing a dict of the query params into urlencode

hughrun replied:
whoops how embarrasing. I meant to clean this up and forgot about it: the plan for this whole PR was "I wonder if this would even work? Let's find out!" hence the trail of mess I left behind me 🤣

hughrun replied:
LOL actually there is an even easier way because `requests` takes a params dict and url encodes the values.

mouse-reeve reviewed (changes requested):
I'll properly check out the branch and try it soon, I'm really excited to have this! For now, a few things I noticed which should be quick to fix :)

mouse-reeve reviewed (changes requested):
This is so cool! Most of the results I tried worked super well, but there were a few searches that produced odd results -- I think primarily when there are multiple results for the name in the ISNI database.

<img width="367" alt="Screen Shot 2021-11-17 at 11 44 44 AM" src="https://user-images.githubusercontent.com/1807695/142273752-0bd80ddd-e8a8-4894-abf4-bec3ba53969e.png">
<img width="420" alt="Screen Shot 2021-11-17 at 11 36 10 AM" src="https://user-images.githubusercontent.com/1807695/142273836-4befecba-7edf-4e95-90ad-1b819a478129.png">

For some of these, it seems like ISNI just isn't providing any useful hints in the data, so the solution might be to have a null state for the helper link that sends you to the ISNI record. In other cases, it's finding text that's in an unexpected language, which might just be an unfortunate thing we live with.

hughrun commented:
Yeah as far as I can tell the language will be from whatever OCLC/ISNI decides is the primary record, because most records are aggregates of a heap of duplicates that are essentially "see references".

I can look further into the duplicate ISNIS, there might be some piece of metadata pointing to something indicating multiple records relate to the same person.

mouse-reeve commented:
It may be duplicate records or it may just be unique records with the same name (I imagine there are plenty of people named Dan Brown), and it's too bad that they aren't providing more consistent info on records but what can you do :)

I think as long as it's linking back to the source that's sufficient, that way you aren't just presented with a wall of indistinguishable names

hughrun commented on bookwyrm/utils/isni.py:
```diff
@@ -0,0 +1,144 @@
+    element = root.find(".//responseRecord")
+    personal_name = element.find(".//forename/..")
+    bio = element.find(".//nameTitle")
+    author = {}
```
I might be being super dumb here but how do I get django to stop complaining that I haven't given an `id` key for `activitypub.Author`?

```python
author = models.Author(
  name=name,
  isni=isni,
  viafId=viaf,
  aliases=aliases,
  bio=bio,
  wikipediaLink=wikipedia
)

return author
```

mouse-reeve replied:
You're not being dumb, it's a required field that makes sense in activitypub and less so in this context. If you have the URL it queried, use that as the id, otherwise, use the isni I think. It normally stores the url used to resolve the object, like in: https://bookwyrm.social/author/2601.json

hughrun replied:
might make sense to use the isni.org url, then.

mouse-reeve replied:
yeah, the isni.org url makes sense. the `id` field gets saved as `origin_id` in the model, which can be helpful for tracking down how data got where it is, but isn't actively used for anything

hughrun replied:
Ok I think I mostly understand this and I have a working update, except for one major bummer. If I do this:

```python
author = author_object.to_model(
  model=models.Author, 
  overwrite=False
)
```
...then I don't seem to have any control over which existing author Bookwyrm uses to "merge". In one case I want to merge with any existing author if the ID matches, and in another case we will never want to merge and always want to create a new author. Currently what's happening is it always merges on a matching name, regardless of anything else.

I'm not sure how I can do either of those things: there doesn't seem to be an obvious way to create a new author in the DB, from an activitypub dataclass, nor to force a new author from the dataclass. But you mentioned above that if I do _something_ then the id automagically becomes the `origin_id` so I guess I need to know what that something is?

Apologies, I'm not familiar with dataclasses and I'm struggling to understand the activitypub code. I can see the sense in using the code that's already there to keep things clean and DRY, but I'm missing some key aspect to making this work.

mouse-reeve replied:
Two things to fix this, first is, in `bookwyrm/models/author.py` on line 30, remove `deduplication_field=True` from the `name` field and run a migration, that will stop it merged based on author name -- this was a better-than-nothing quick fix for openlibrary issues that I meant to remove ages ago. To create a new author, you could create an `Author` model object and pass it in with the instance field. If that doesn't work I can modify the `to_model` function to specify requiring a new object.

You shouldn't need to do anything to get the origin_id field set, that happens in the model save function: https://github.com/bookwyrm-social/bookwyrm/blob/main/bookwyrm/models/book.py#L65 if the object is being created.

Truly no apologies necessary, this is the dark underbelly of the code that makes sense only inside my own head. I really appreciate you bearing with me on it, I know it's very unintuitive and does a lot of hand-wavey surprise things in non-obvious places.

hughrun replied:
yesss this rules.

Django doesn't seem to want/need a migration after removing name as a dedupe field. Works like a charm as described, it's a good way to do this, and now I get it. Thanks!

hughrun commented:
Ok I have cleaned this up, thanks for the tips @mouse-reeve !

- now uses activitypub Author dataclass instead of bespoke dict
- isni api query as a timeout
- author hint is now always "Author of [title]", a with a fallback in the very unlikely event there's no titles listed in isni
- title query for above prefers English-speaking sources
- fixed a couple of little bugs in my original PR code

mouse-reeve reviewed (approved):
Yay!
